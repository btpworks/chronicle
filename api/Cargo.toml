[package]
name = "api"
version = "0.1.0"
edition = "2021"
build = "build.rs"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
custom_error = "1.9.2"
diesel_migrations = { version = "^2.0.0", features= ["sqlite"], git = "https://github.com/diesel-rs/diesel/"}
diesel = { version = "^2.0.0", features = ["sqlite", "uuid", "chrono"], git = "https://github.com/diesel-rs/diesel" }
dotenv = "0.15.0"
iref = "2.0.3"
iref-enum = "1.2.0"
json = "0.12.4"
json-ld = "0.4.0"
k256 = { version = "0.9.6", features = ["pem","std"] }
pkcs8 = { version = "0.7.6", features = ["pem","std","alloc"] }
static-iref = "1.0.0"
user-error = "1.2.8"
uuid = { version = "0.8.2", features = ["v4"] }
validator = { version = "0.14.0", features = ["derive"] }
url = "2.2.2"
glob = "0.3.0"
proto = {path="../sawtooth_protocol"}
common = {path="../common"}
tracing = "0.1.29"
derivative = "2.2.0"
chrono = "0.4.19"
multimap = "0.8.3"
hex = "0.4.3"
async-change-tracker = "0.3.1"
bui-backend = "0.12.0"
bui-backend-types = "0.8.0"
tokio = { version = "1.14.0", features = ["time", "macros", "rt-multi-thread"] }
parking_lot = "0.11.2"
includedir = "0.6.0"
phf = "0.8"
serde = "1.0.130"
serde_derive = "1.0.130"
futures = "0.3"
async-trait = "0.1.51"
crossbeam = "0.8.1"
open = "2.0.1"


[build-dependencies]
bui-backend-codegen = "0.9.0"
bts = "0.3.0"
btl = "0.3.0"

[dev-dependencies]
insta = "1.8.0"
tempfile = "3.2.0"
tracing-subscriber = "0.3.1"
trunk = "0.14.0"


[features]
# Bundle files into executable
bundle_files = ["bui-backend/bundle_files", "bui-backend-codegen/bundle_files"]
# Serve files from disk at runtime
serve_files = ["bui-backend/serve_files", "bui-backend-codegen/serve_files"]

