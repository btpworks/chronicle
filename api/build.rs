fn main() -> Result<(), std::io::Error> {
    use btl::{cd, shell};
    use std::{env, path::PathBuf};

    //let release = env::var("PROFILE").expect("expected PROFILE to be set by Cargo") != "debug";
    //let out_dir = PathBuf::from(env::var("OUT_DIR").expect("expected OUT_DIR to be set by Cargo"));

    let files_dir: std::path::PathBuf = [PathBuf::from("../web-target")].iter().collect();
    bui_backend_codegen::codegen(&files_dir, "public.rs").expect("codegen failed");

    Ok(())
}
