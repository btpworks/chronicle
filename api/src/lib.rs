#[macro_use]
extern crate diesel;
#[macro_use]
extern crate diesel_migrations;
#[macro_use]
extern crate iref_enum;
#[macro_use]
extern crate json;

mod models;
mod schema;
mod vocab;

use async_std::task;
use custom_error::*;
use derivative::*;
use diesel::{prelude::*, sqlite::SqliteConnection};
use json::JsonValue;
use json_ld::{context::Local, Document, JsonContext, NoLoader};
use std::path::Path;

use common::{
    models::{ChronicleTransaction, CreateAgent, CreateNamespace},
    signing::{DirectoryStoredKeys, SignerError},
};
use iref::IriBuf;
use models::Agent;
use proto::messaging::{SawtoothValidator, SubmissionError};
use tracing::instrument;
use url::Url;
use user_error::UFE;
use uuid::Uuid;

embed_migrations!();

custom_error! {pub ApiError
    Db{source: diesel::result::Error}                           = "Database operation failed",
    DbConnection{source: diesel::ConnectionError}               = "Database connection failed",
    DbMigration{source: diesel_migrations::RunMigrationsError}  = "Database migration failed",
    Iri{source: iref::Error}                                    = "Invalid IRI",
    JsonLD{source: json_ld::Error}                              = "Json LD processing",
    Api{source: SubmissionError}                                = "Json LD processing",
    Signing{source: SignerError}                                = "Signing",
}

impl UFE for ApiError {}

#[derive(Debug)]
pub enum NamespaceCommand {
    Create { name: String },
}

#[derive(Debug)]
pub enum AgentCommand {
    Create {
        name: String,
        namespace: String,
    },
    RegisterKey {
        name: String,
        namespace: String,
        public: String,
        private: Option<String>,
    },
}

#[derive(Debug)]
pub enum ApiCommand {
    NameSpace(NamespaceCommand),
    Agent(AgentCommand),
}

#[derive(Debug)]
pub enum ApiResponse {
    Unit,
    Iri(IriBuf),
    Document(JsonValue),
}

#[derive(Derivative)]
#[derivative(Debug)]
pub struct Api {
    #[derivative(Debug = "ignore")]
    connection: SqliteConnection,
    ledger: SawtoothValidator,
}

impl Api {
    pub fn new(
        database_url: &str,
        sawtooth_url: &Url,
        secret_path: &Path,
    ) -> Result<Self, ApiError> {
        let connection = SqliteConnection::establish(database_url)?;
        embedded_migrations::run(&connection)?;

        let ledger = SawtoothValidator::new(
            sawtooth_url,
            DirectoryStoredKeys::new(secret_path)?.default(),
        );
        Ok(Api { connection, ledger })
    }

    #[instrument]
    fn create_namespace(&self, name: &str) -> Result<ApiResponse, ApiError> {
        let uuid = Uuid::new_v4();
        let newnamespace = models::NewNamespace {
            name,
            uuid: &uuid.to_string(),
        };

        diesel::insert_or_ignore_into(schema::namespace::table)
            .values(&newnamespace)
            .execute(&self.connection)
            .map(|_| {
                self.ledger
                    .submit(vec![ChronicleTransaction::CreateNamespace(
                        CreateNamespace {
                            id: IriBuf::from(&newnamespace).as_iri().into(),
                        },
                    )])
            })
            .map(|_| ApiResponse::Unit)
            .map_err(ApiError::from)
    }

    #[instrument]
    fn create_agent(&self, name: &str, namespace: &str) -> Result<ApiResponse, ApiError> {
        use self::schema::namespace::dsl as ns;

        self.create_namespace(namespace)?;

        let namespace_data = ns::namespace
            .filter(ns::name.eq(namespace))
            .first::<models::NameSpace>(&self.connection)?;

        let namespace_iri = IriBuf::new(&format!(
            "chronicle:{}/{}#",
            namespace_data.name, namespace_data.uuid
        ))?;

        let id = IriBuf::new(&format!("{}agent:{}", namespace_iri, name))?;

        let input = object! {
            "http://www.w3.org/ns/prov#Agent" : {
                "@id": (id.as_str())
            }
        };
        let context = object! {};

        let processed_context =
            task::block_on(context.process::<JsonContext, _>(&mut NoLoader, None))?;

        let output = task::block_on(input.compact(&processed_context, &mut NoLoader))?;

        let newagent = models::NewAgent {
            name,
            namespace,
            current: 0,
            publickey: None,
            privatekeypath: None,
        };

        diesel::insert_or_ignore_into(schema::agent::table)
            .values(&newagent)
            .execute(&self.connection)
            .map(|_| {
                self.ledger
                    .submit(vec![ChronicleTransaction::CreateAgent(CreateAgent {
                        id: IriBuf::from(&newagent).as_iri().into(),
                        namespace: IriBuf::from(&namespace_data).as_iri().into(),
                    })])
            })
            .map_err(ApiError::from)
            .map(|_| ApiResponse::Document(output))
    }

    #[instrument]
    pub fn dispatch(&self, command: ApiCommand) -> Result<ApiResponse, ApiError> {
        match command {
            ApiCommand::NameSpace(NamespaceCommand::Create { name }) => {
                self.create_namespace(&name)
            }
            ApiCommand::Agent(AgentCommand::Create { name, namespace }) => {
                self.create_agent(&name, &namespace)
            }
            ApiCommand::Agent(AgentCommand::RegisterKey {
                name,
                namespace,
                public,
                private,
            }) => self.register_key(&name, &namespace, &public, private),
        }
    }

    #[instrument]
    fn get_agent(
        &self,
        name: &str,
        namespace: &str,
    ) -> Result<Option<Agent>, diesel::result::Error> {
        use self::schema::agent::dsl as ns;
        ns::agent
            .filter(ns::name.eq(name).and(ns::namespace.eq(namespace)))
            .first::<models::Agent>(&self.connection)
            .optional()
    }

    #[instrument]
    fn register_key(
        &self,
        name: &str,
        namespace: &str,
        publickey: &str,
        privatekeypath: Option<String>,
    ) -> Result<ApiResponse, ApiError> {
        use self::schema::agent::dsl as ns;

        if let Some(Agent { current, .. }) = self.get_agent(name, namespace)? {
            let update = models::NewAgent {
                name,
                namespace,
                current,
                publickey: Some(publickey),
                privatekeypath: privatekeypath.as_deref(),
            };

            diesel::update(schema::agent::table)
                .filter(ns::name.eq(namespace).and(ns::namespace.eq(namespace)))
                .set(update)
                .execute(&self.connection)
                .map_err(ApiError::from)
                .map(|_| ApiResponse::Unit)
        } else {
            let insert = models::NewAgent {
                name,
                namespace,
                publickey: Some(publickey),
                privatekeypath: privatekeypath.as_deref(),
                ..Default::default()
            };

            diesel::insert_into(schema::agent::table)
                .values(insert)
                .execute(&self.connection)
                .map_err(ApiError::from)
                .map(|_| ApiResponse::Unit)
        }
    }
}
