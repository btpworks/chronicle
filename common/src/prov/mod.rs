mod id;
mod models;
pub mod transactions;
pub mod vocab;

pub use id::*;
pub use models::*;
