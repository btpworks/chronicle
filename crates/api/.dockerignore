sqlite_test
db.sqlite
