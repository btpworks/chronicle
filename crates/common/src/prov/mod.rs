mod id;
mod model;
pub mod operations;
pub mod vocab;

pub use id::*;
pub use model::*;
