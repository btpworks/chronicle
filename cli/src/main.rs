//! The default graphql api - only abstract resources
//! We delegate to the underlying concrete, attributeless graphql objects in the api crate,
//! wrapping those in our types here.
#![cfg_attr(feature = "strict", deny(warnings))]
use api::{
    self,
    chronicle_graphql::{self, ChronicleGraphQl, Evidence, Identity, Namespace, Submission},
};
use async_graphql::{
    connection::{Connection, EmptyFields},
    *,
};
use bootstrap::*;
use chrono::{DateTime, Utc};
use common::prov::{vocab::Prov, ActivityId, AgentId, DomaintypeId, EntityId};
use iref::Iri;

#[derive(Default, InputObject)]
pub struct Attributes {
    #[graphql(name = "type")]
    pub typ: Option<String>,
}

impl From<Attributes> for common::attributes::Attributes {
    fn from(attributes: Attributes) -> Self {
        common::attributes::Attributes {
            typ: attributes.typ.map(|typ| DomaintypeId::from_name(&typ)),
            ..Default::default()
        }
    }
}

pub struct Activity(chronicle_graphql::Activity);

#[Object]
impl Activity {
    async fn id(&self) -> ActivityId {
        ActivityId::from_name(&*self.0.name)
    }

    async fn namespace<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Namespace> {
        chronicle_graphql::activity::namespace(self.0.namespace_id, ctx).await
    }

    async fn name(&self) -> &str {
        &self.0.name
    }

    async fn started(&self) -> Option<DateTime<Utc>> {
        self.0.started.map(|x| DateTime::from_utc(x, Utc))
    }

    async fn ended(&self) -> Option<DateTime<Utc>> {
        self.0.ended.map(|x| DateTime::from_utc(x, Utc))
    }

    #[graphql(name = "type")]
    async fn typ(&self) -> String {
        if let Some(ref typ) = self.0.domaintype {
            typ.to_string()
        } else {
            Iri::from(Prov::Activity).to_string()
        }
    }

    async fn was_associated_with<'a>(
        &self,
        ctx: &Context<'a>,
    ) -> async_graphql::Result<Vec<Agent>> {
        Ok(
            chronicle_graphql::activity::was_associated_with(self.0.id, ctx)
                .await?
                .into_iter()
                .map(Agent)
                .collect(),
        )
    }

    async fn used<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Vec<Entity>> {
        Ok(chronicle_graphql::activity::used(self.0.id, ctx)
            .await?
            .into_iter()
            .map(Entity)
            .collect())
    }
}
pub struct Entity(chronicle_graphql::Entity);

#[Object]
impl Entity {
    async fn id(&self) -> EntityId {
        EntityId::from_name(&*self.0.name)
    }

    async fn namespace<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Namespace> {
        chronicle_graphql::entity::namespace(self.0.namespace_id, ctx).await
    }

    async fn name(&self) -> &str {
        &self.0.name
    }

    #[graphql(name = "type")]
    async fn typ(&self) -> String {
        if let Some(ref typ) = self.0.domaintype {
            typ.to_string()
        } else {
            Iri::from(Prov::Agent).to_string()
        }
    }

    async fn evidence<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Option<Evidence>> {
        chronicle_graphql::entity::evidence(self.0.attachment_id, ctx).await
    }

    async fn was_generated_by<'a>(
        &self,
        ctx: &Context<'a>,
    ) -> async_graphql::Result<Vec<Activity>> {
        Ok(chronicle_graphql::entity::was_generated_by(self.0.id, ctx)
            .await?
            .into_iter()
            .map(Activity)
            .collect())
    }

    async fn was_derived_from<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Vec<Entity>> {
        Ok(chronicle_graphql::entity::was_derived_from(self.0.id, ctx)
            .await?
            .into_iter()
            .map(Entity)
            .collect())
    }

    async fn had_primary_source<'a>(
        &self,
        ctx: &Context<'a>,
    ) -> async_graphql::Result<Vec<Entity>> {
        Ok(
            chronicle_graphql::entity::had_primary_source(self.0.id, ctx)
                .await?
                .into_iter()
                .map(Entity)
                .collect(),
        )
    }

    async fn was_revision_of<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Vec<Entity>> {
        Ok(chronicle_graphql::entity::was_revision_of(self.0.id, ctx)
            .await?
            .into_iter()
            .map(Entity)
            .collect())
    }

    async fn was_quoted_from<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Vec<Entity>> {
        Ok(chronicle_graphql::entity::was_quoted_from(self.0.id, ctx)
            .await?
            .into_iter()
            .map(Entity)
            .collect())
    }
}

pub struct Agent(chronicle_graphql::Agent);

#[Object]
impl Agent {
    async fn id(&self) -> AgentId {
        AgentId::from_name(&*self.0.name)
    }

    async fn name(&self) -> &str {
        &self.0.name
    }

    async fn namespace<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Namespace> {
        chronicle_graphql::agent::namespace(self.0.namespace_id, ctx).await
    }

    async fn identity<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Option<Identity>> {
        chronicle_graphql::agent::identity(self.0.identity_id, ctx).await
    }

    async fn acted_on_behalf_of<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Vec<Agent>> {
        Ok(chronicle_graphql::agent::acted_on_behalf_of(self.0.id, ctx)
            .await?
            .into_iter()
            .map(Self)
            .collect())
    }

    #[graphql(name = "type")]
    async fn typ(&self) -> String {
        Iri::from(Prov::Agent).to_string()
    }
}

#[derive(Copy, Clone)]
pub struct Mutation;

#[Object]
impl Mutation {
    pub async fn agent<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: Option<String>,
        attributes: Attributes,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::agent(ctx, name, namespace, attributes.into()).await
    }

    pub async fn activity<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: Option<String>,
        attributes: Attributes,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::activity(ctx, name, namespace, attributes.into()).await
    }

    pub async fn entity<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: Option<String>,
        attributes: Attributes,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::entity(ctx, name, namespace, attributes.into()).await
    }

    pub async fn acted_on_behalf_of<'a>(
        &self,
        ctx: &Context<'a>,
        namespace: Option<String>,
        responsible: AgentId,
        delegate: AgentId,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::acted_on_behalf_of(ctx, namespace, responsible, delegate).await
    }

    pub async fn was_derived_from<'a>(
        &self,
        ctx: &Context<'a>,
        namespace: Option<String>,
        generated_entity: EntityId,
        used_entity: EntityId,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::was_derived_from(ctx, namespace, generated_entity, used_entity)
            .await
    }

    pub async fn was_revision_of<'a>(
        &self,
        ctx: &Context<'a>,
        namespace: Option<String>,
        generated_entity: EntityId,
        used_entity: EntityId,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::was_revision_of(ctx, namespace, generated_entity, used_entity)
            .await
    }

    pub async fn had_primary_source<'a>(
        &self,
        ctx: &Context<'a>,
        namespace: Option<String>,
        generated_entity: EntityId,
        used_entity: EntityId,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::had_primary_source(
            ctx,
            namespace,
            generated_entity,
            used_entity,
        )
        .await
    }

    pub async fn was_quoted_from<'a>(
        &self,
        ctx: &Context<'a>,
        namespace: Option<String>,
        generated_entity: EntityId,
        used_entity: EntityId,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::was_quoted_from(ctx, namespace, generated_entity, used_entity)
            .await
    }

    pub async fn generate_key<'a>(
        &self,
        ctx: &Context<'a>,
        id: AgentId,
        namespace: Option<String>,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::generate_key(ctx, id, namespace).await
    }

    pub async fn start_activity<'a>(
        &self,
        ctx: &Context<'a>,
        id: ActivityId,
        namespace: Option<String>,
        agent: AgentId,
        time: Option<DateTime<Utc>>,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::start_activity(ctx, id, namespace, agent, time).await
    }

    pub async fn end_activity<'a>(
        &self,
        ctx: &Context<'a>,
        id: ActivityId,
        namespace: Option<String>,
        agent: AgentId,
        time: Option<DateTime<Utc>>,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::end_activity(ctx, id, namespace, agent, time).await
    }

    pub async fn used<'a>(
        &self,
        ctx: &Context<'a>,
        activity: ActivityId,
        id: EntityId,
        namespace: Option<String>,
        _typ: Option<String>,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::used(ctx, activity, id, namespace).await
    }

    pub async fn was_generated_by<'a>(
        &self,
        ctx: &Context<'a>,
        activity: ActivityId,
        id: EntityId,
        namespace: Option<String>,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::was_generated_by(ctx, activity, id, namespace).await
    }

    pub async fn has_attachment<'a>(
        &self,
        ctx: &Context<'a>,
        id: EntityId,
        namespace: Option<String>,
        attachment: Upload,
        agent: AgentId,
        locator: String,
    ) -> async_graphql::Result<Submission> {
        chronicle_graphql::mutation::has_attachment(ctx, id, namespace, attachment, agent, locator)
            .await
    }
}

#[derive(Copy, Clone)]
pub struct Query;

#[Object]
impl Query {
    #[allow(clippy::too_many_arguments)]
    pub async fn agents_by_type<'a>(
        &self,
        ctx: &Context<'a>,
        agent_type: ID,
        namespace: Option<ID>,
        after: Option<String>,
        before: Option<String>,
        first: Option<i32>,
        last: Option<i32>,
    ) -> async_graphql::Result<Connection<i32, Agent, EmptyFields, EmptyFields>> {
        Ok(chronicle_graphql::query::agents_by_type(
            ctx, agent_type, namespace, after, before, first, last,
        )
        .await?
        .map_node(Agent))
    }

    pub async fn agent_by_id<'a>(
        &self,
        ctx: &Context<'a>,
        id: AgentId,
        namespace: Option<String>,
    ) -> async_graphql::Result<Option<Agent>> {
        Ok(chronicle_graphql::query::agent_by_id(ctx, id, namespace)
            .await?
            .map(Agent))
    }

    pub async fn entity_by_id<'a>(
        &self,
        ctx: &Context<'a>,
        id: EntityId,
        namespace: Option<String>,
    ) -> async_graphql::Result<Option<Entity>> {
        Ok(chronicle_graphql::query::entity_by_id(ctx, id, namespace)
            .await?
            .map(Entity))
    }
}

#[tokio::main]
pub async fn main() {
    bootstrap(ChronicleGraphQl::new(Query, Mutation)).await
}

#[cfg(test)]
mod test {
    use super::{Mutation, Query};
    use crate::chronicle_graphql::{Store, Subscription};
    use api::{Api, ConnectionOptions, UuidGen};
    use async_graphql::{Request, Schema};
    use common::ledger::InMemLedger;
    use diesel::{
        r2d2::{ConnectionManager, Pool},
        SqliteConnection,
    };
    use std::time::Duration;
    use tempfile::TempDir;
    use tracing::Level;
    use uuid::Uuid;

    #[derive(Debug, Clone)]
    struct SameUuid;

    impl UuidGen for SameUuid {
        fn uuid() -> Uuid {
            Uuid::parse_str("5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea").unwrap()
        }
    }

    async fn test_schema() -> Schema<Query, Mutation, Subscription> {
        tracing_log::LogTracer::init_with_filter(tracing::log::LevelFilter::Trace).ok();
        tracing_subscriber::fmt()
            .pretty()
            .with_max_level(Level::TRACE)
            .try_init()
            .ok();

        let secretpath = TempDir::new().unwrap();

        // We need to use a real file for sqlite, as in mem either re-creates between
        // macos temp dir permissions don't work with sqlite
        std::fs::create_dir("./sqlite_test").ok();
        let dbid = Uuid::new_v4();
        let mut ledger = InMemLedger::new();
        let reader = ledger.reader();

        let pool = Pool::builder()
            .connection_customizer(Box::new(ConnectionOptions {
                enable_wal: true,
                enable_foreign_keys: true,
                busy_timeout: Some(Duration::from_secs(2)),
            }))
            .build(ConnectionManager::<SqliteConnection>::new(&*format!(
                "./sqlite_test/db{}.sqlite",
                dbid
            )))
            .unwrap();

        let dispatch = Api::new(
            pool.clone(),
            ledger,
            reader,
            &secretpath.into_path(),
            SameUuid,
        )
        .await
        .unwrap();

        Schema::build(Query, Mutation, Subscription)
            .data(Store::new(pool))
            .data(dispatch)
            .finish()
    }

    #[tokio::test]
    async fn delegation() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                actedOnBehalfOf(
                    responsible: "http://blockchaintp.com/chronicle/ns#agent:responsible",
                    delegate: "http://blockchaintp.com/chronicle/ns#agent:delegate",
                    ) {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;

        let derived = schema
            .execute(Request::new(
                r#"
            query {
                agentById(id: "http://blockchaintp.com/chronicle/ns#agent:responsible") {
                    id
                    name
                    actedOnBehalfOf {
                        id
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_toml_snapshot!(derived);
    }

    #[tokio::test]
    async fn untyped_derivation() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                wasDerivedFrom(generatedEntity: "http://blockchaintp.com/chronicle/ns#entity:generated",
                               usedEntity: "http://blockchaintp.com/chronicle/ns#entity:used") {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;
        let derived = schema
            .execute(Request::new(
                r#"
            query {
                entityById(id: "http://blockchaintp.com/chronicle/ns#entity:generated") {
                    id
                    name
                    wasDerivedFrom {
                        id
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_toml_snapshot!(derived);
    }

    #[tokio::test]
    async fn primary_source() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                hadPrimarySource(generatedEntity: "http://blockchaintp.com/chronicle/ns#entity:generated",
                               usedEntity: "http://blockchaintp.com/chronicle/ns#entity:used") {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;
        let derived = schema
            .execute(Request::new(
                r#"
            query {
                entityById(id: "http://blockchaintp.com/chronicle/ns#entity:generated") {
                    id
                    name
                    wasDerivedFrom {
                        id
                    }
                    hadPrimarySource {
                        id
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_toml_snapshot!(derived);
    }

    #[tokio::test]
    async fn revision() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                wasRevisionOf(generatedEntity: "http://blockchaintp.com/chronicle/ns#entity:generated",
                            usedEntity: "http://blockchaintp.com/chronicle/ns#entity:used") {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;
        let derived = schema
            .execute(Request::new(
                r#"
            query {
                entityById(id: "http://blockchaintp.com/chronicle/ns#entity:generated") {
                    id
                    name
                    wasDerivedFrom {
                        id
                    }
                    wasRevisionOf {
                        id
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_toml_snapshot!(derived);
    }

    #[tokio::test]
    async fn quotation() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                wasQuotedFrom(generatedEntity: "http://blockchaintp.com/chronicle/ns#entity:generated",
                            usedEntity: "http://blockchaintp.com/chronicle/ns#entity:used") {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;
        let derived = schema
            .execute(Request::new(
                r#"
            query {
                entityById(id: "http://blockchaintp.com/chronicle/ns#entity:generated") {
                    id
                    name
                    wasDerivedFrom {
                        id
                    }
                    wasQuotedFrom {
                        id
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_toml_snapshot!(derived);
    }

    #[tokio::test]
    async fn agent_can_be_created() {
        let schema = test_schema().await;

        let create = schema
            .execute(Request::new(
                r#"
            mutation {
                agent(name:"bobross", attributes: { type: "artist" }) {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(create);
    }

    #[tokio::test]
    async fn query_agents_by_cursor() {
        let schema = test_schema().await;

        for i in 0..100 {
            schema
                .execute(Request::new(format!(
                    r#"
            mutation {{
                agent(name:"bobross{}", attributes: {{ type: "artist"}}) {{
                    context
                }}
            }}
        "#,
                    i
                )))
                .await;
        }
        tokio::time::sleep(Duration::from_millis(100)).await;

        let default_cursor = schema
            .execute(Request::new(
                r#"
                query {
                agentsByType(agentType: "artist") {
                    pageInfo {
                        hasPreviousPage
                        hasNextPage
                        startCursor
                        endCursor
                    }
                    edges {
                        node {
                            id,
                            name
                        }
                        cursor
                    }
                }
                }
        "#,
            ))
            .await;

        insta::assert_json_snapshot!(default_cursor);

        let middle_cursor = schema
            .execute(Request::new(
                r#"
                query {
                agentsByType(agentType: "artist", first: 20, after: "3") {
                    pageInfo {
                        hasPreviousPage
                        hasNextPage
                        startCursor
                        endCursor
                    }
                    edges {
                        node {
                            id,
                            name
                        }
                        cursor
                    }
                }
                }
        "#,
            ))
            .await;

        insta::assert_json_snapshot!(middle_cursor);

        let out_of_bound_cursor = schema
            .execute(Request::new(
                r#"
                query {
                agentsByType(agentType: "artist", first: 20, after: "90") {
                    pageInfo {
                        hasPreviousPage
                        hasNextPage
                        startCursor
                        endCursor
                    }
                    edges {
                        node {
                            id,
                            name
                        }
                        cursor
                    }
                }
                }
        "#,
            ))
            .await;

        insta::assert_json_snapshot!(out_of_bound_cursor);
    }
}
