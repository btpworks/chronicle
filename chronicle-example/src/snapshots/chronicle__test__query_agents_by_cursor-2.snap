---
source: cli/src/main.rs
assertion_line: 759
expression: middle_cursor
---
{
  "data": {
    "agentsByType": {
      "pageInfo": {
        "hasPreviousPage": true,
        "hasNextPage": true,
        "startCursor": "4",
        "endCursor": "23"
      },
      "edges": [
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross12",
            "name": "bobross12"
          },
          "cursor": "4"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross13",
            "name": "bobross13"
          },
          "cursor": "5"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross14",
            "name": "bobross14"
          },
          "cursor": "6"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross15",
            "name": "bobross15"
          },
          "cursor": "7"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross16",
            "name": "bobross16"
          },
          "cursor": "8"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross17",
            "name": "bobross17"
          },
          "cursor": "9"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross18",
            "name": "bobross18"
          },
          "cursor": "10"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross19",
            "name": "bobross19"
          },
          "cursor": "11"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross2",
            "name": "bobross2"
          },
          "cursor": "12"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross20",
            "name": "bobross20"
          },
          "cursor": "13"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross21",
            "name": "bobross21"
          },
          "cursor": "14"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross22",
            "name": "bobross22"
          },
          "cursor": "15"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross23",
            "name": "bobross23"
          },
          "cursor": "16"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross24",
            "name": "bobross24"
          },
          "cursor": "17"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross25",
            "name": "bobross25"
          },
          "cursor": "18"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross26",
            "name": "bobross26"
          },
          "cursor": "19"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross27",
            "name": "bobross27"
          },
          "cursor": "20"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross28",
            "name": "bobross28"
          },
          "cursor": "21"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross29",
            "name": "bobross29"
          },
          "cursor": "22"
        },
        {
          "node": {
            "id": "http://blockchaintp.com/chronicle/ns#agent:bobross3",
            "name": "bobross3"
          },
          "cursor": "23"
        }
      ]
    }
  }
}
